packages/thirdweb/src/wallets/in-app/core/authentication/passkeys.ts

```diff
@@ -0,0 +1,209 @@
+import { getClientFetch } from "../../../../utils/fetch.js";
+import type { AsyncStorage } from "../../../../utils/storage/AsyncStorage.js";
+import type { Ecosystem } from "../../web/types.js";
+import { ClientScopedStorage } from "./client-scoped-storage.js";
+import type { AuthStoredTokenWithCookieReturnType } from "./types.js";
+
+function getVerificationPath() {
+  return `${getThirdwebBaseUrl(
+    "inAppWallet",
+  )}/api/2024-05-05/login/passkey/callback`;
+}
+function getChallengePath(type: "sign-in" | "sign-up", username?: string) {
+  return `${getThirdwebBaseUrl(
+    "inAppWallet",
+  )}/api/2024-05-05/login/passkey?type=${type}${
+    username ? `&username=${username}` : ""
+  }`;
+}
+
+export type RegisterResult = {
+  authenticatorData: string;
+  credentialId: string;
+  clientData: string;
+  credential: {
+    publicKey: string;
+    algorithm: string;
+  };
+  origin: string;
+};
+
+export type AuthenticateResult = {
+  credentialId: string;
+  authenticatorData: string;
+  clientData: string;
+  signature: string;
+  origin: string;
+};
+
+export type RpInfo = { name: string; id: string };
+
+export interface PasskeyClient {
+  register: (args: {
+    name: string;
+    challenge: string;
+    rp: RpInfo;
+  }) => Promise<RegisterResult>;
+  authenticate: (args: {
+    credentialId: string | undefined;
+    challenge: string;
+    rp: RpInfo;
+  }) => Promise<AuthenticateResult>;
+  isAvailable: () => boolean;
+}
+
+export async function registerPasskey(options: {
+  client: ThirdwebClient;
+  storage: AsyncStorage;
+  passkeyClient: PasskeyClient;
+  ecosystem?: Ecosystem;
+  username?: string;
+  rp: RpInfo;
+}): Promise<AuthStoredTokenWithCookieReturnType> {
+  if (!options.passkeyClient.isAvailable()) {
+    throw new Error("Passkeys are not available on this device");
+  }
+  const storage = new ClientScopedStorage({
+    storage: options.storage,
+    clientId: options.client.clientId,
+    ecosystemId: options.ecosystem?.id,
+  });
+  const fetchWithId = getClientFetch(options.client, options.ecosystem);
+  const generatedName = options.username ?? generateUsername(options.ecosystem);
+  // 1. request challenge from  server
+  const res = await fetchWithId(getChallengePath("sign-up", generatedName));
+  const challengeData = await res.json();
+  if (!challengeData.challenge) {
+    throw new Error("No challenge received");
+  }
+  const challenge = challengeData.challenge;
+
+  // 2. initiate registration
+  const registration = await options.passkeyClient.register({
+    name: generatedName,
+    challenge,
+    rp: options.rp,
+  });
+
+  const customHeaders: Record<string, string> = {};
+  if (options.ecosystem?.partnerId) {
+    customHeaders["x-ecosystem-partner-id"] = options.ecosystem.partnerId;
+  }
+  if (options.ecosystem?.id) {
+    customHeaders["x-ecosystem-id"] = options.ecosystem.id;
+  }
+
+  // 3. send the registration object to the server
+  const verifRes = await fetchWithId(getVerificationPath(), {
```

Comment:
Are you not able to use `getClientFetch` here? It adds the ecosystem headers for you